### PR TITLE
Issue 765 - Addressed LocationLevel Catalog requiring start date

### LINKED ISSUE #765: Getting location level list requires knowing valid begin time
A user recently tried to list location levels matching a mask and received no results. Changing begin to 1900-01-01 provided all appropriate results, but doesn't seem the behavior we desire.

Either the simple GET endpoint of levels should act like the catalog giving the requestor the known results, or we should add a new dataset to the /catalog endpoint.

### PR DESCRIPTION
**Pull request by zack-rma**
Fixes [#765](https://github.com/USACE/cwms-data-api/issues/765) - Adjusted start date default to 1900

### CODE REVIEW

File: cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java [hunk 1]
```diff
@@ -295,7 +294,7 @@ public void getAll(@NotNull Context ctx) {
                 if (begin != null) {
                     beginZdt = DateUtils.parseUserDate(begin, timezone);
                 } else {
-                    beginZdt = endZdt.minusHours(24);
+                    beginZdt = ZonedDateTime.of(1900, 1, 1, 0, 0, 0, 0, tz);
```
Comment (MikeNeilson) [inline cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java]: While generally everything practical does start at 1900, it would probably make sense to set this as far back as ZonedDateTime allows. Does ZonedDateTime have an "earliest date" function?

**Review by MikeNeilson — CHANGES_REQUESTED** [review]

File: cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java (hunk 1 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java]: I looked into this a bit. ZonedDateTime does not appear to have an earliest time function, but I tried generating an earliest time using ZonedDateTime from Instant.MIN. Unfortunately, this is outside of the date values supported, so I went with the earliest supported Instant EpochMilli of -999999999.

Comment (MikeNeilson) [reply cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java]: Works for me. Thanks.

**Review by MikeNeilson — DISMISSED** [review]

**Comment by zack-rma** [comment]
Rebased to fix build errors

File: cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java [hunk 2]
```diff
@@ -295,7 +295,8 @@ public void getAll(@NotNull Context ctx) {
                 if (begin != null) {
                     beginZdt = DateUtils.parseUserDate(begin, timezone);
                 } else {
-                    beginZdt = endZdt.minusHours(24);
+                    // uses earliest supported time in CWMS
+                    beginZdt = ZonedDateTime.ofInstant(Instant.ofEpochMilli(-999999999L), tz);
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java]: this instant is `1969-12-20T10:13:20.001Z` which is definitely not the earliest supported time in CWMS

File: cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java [hunk 3]
```diff
@@ -220,8 +220,8 @@ public void delete(@NotNull Context ctx, @NotNull String levelId) {
                         + "specified or default units above the NGVD-29 datum."),
                 @OpenApiParam(name = BEGIN, description = "Specifies the start of the time "
                         + "window for data to be included in the response. If this field is "
-                        + "not specified, any required time window begins 24 hours prior to "
-                        + "the specified or default end time."),
+                        + "not specified, any required time window begins 1/1/1900 at "
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java]: This date doesn't match up with what is actually used.

**Review by adamkorynta — CHANGES_REQUESTED** [review]

File: cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java (hunk 2 above)
Comment (adamkorynta) [reply cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java]: Tests should be added to test edge case of the earliest value

File: cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java (hunk 3 above)
Comment (adamkorynta) [reply cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java]: I would think leaving off the beginning or ending empty would just remove them from the select query bounds rather filling in a default.

File: cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java (hunk 2 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java]: I realized the instant value I used is actually the minimum year allowed by CWMS.

I added a test case for the minimum time allowed by the database, which is much more recent than the CWMS allowed value.

File: cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java (hunk 3 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java]: I updated the description to better match actual behavior. I also changed the default values to null, which causes the DAO to remove them from the query.

File: cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java [hunk 4]
```diff
@@ -280,23 +278,14 @@ public void getAll(@NotNull Context ctx) {
...
-
-                ZonedDateTime endZdt = end != null ? DateUtils.parseUserDate(end, timezone) :
-                        ZonedDateTime.now(tz);
-                ZonedDateTime beginZdt;
-                if (begin != null) {
-                    beginZdt = DateUtils.parseUserDate(begin, timezone);
-                } else {
-                    beginZdt = endZdt.minusHours(24);
-                }
+                ZonedDateTime endZdt = end != null ? DateUtils.parseUserDate(end, timezone) : null;
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java]: use `requiredInstant(ctx, END);` or `requiredZdt(ctx, END);` (not sure why this time needs to be zoned).

File: cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java [hunk 5]
```diff
@@ -315,12 +304,9 @@ public void getAll(@NotNull Context ctx) {
                 ctx.status(HttpServletResponse.SC_OK);
                 ctx.result(results);
                 requestResultSize.update(results.length());
-                if (isLegacyVersion)
-                {
+                if (isLegacyVersion) {
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java:308]: isn't this always false given the if statement above?

File: cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java (hunk 4 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java]: The beginning and end parameters are not required fields. I replaced the ternaries with `queryParamAsZdt` instead.

File: cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java (hunk 5 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java:308]: Not necessarily. If the format parameter is specified, the statement above will be skipped and the version can still be either legacy or not.

Comment (adamkorynta) [reply cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java:308]: oh I see, it's one of those that uses the format parameter, not just the header

**Review by adamkorynta — APPROVED** [review]
